File: runners/csharp/Kikker.Billing.IntegrationTests/BillingInteractionLogTest.cs
using System.Text.Json;
using Kikker.Billing.IntegrationTests.Fixtures;

namespace Kikker.Billing.IntegrationTests;

[Collection(nameof(BillingFlowCollection))]
public sealed class BillingInteractionLogTest
{
    private readonly BillingFlowFixture fixture;

    public BillingInteractionLogTest(BillingFlowFixture fixture)
    {
        this.fixture = fixture;
    }

    [Fact]
    [Trait("Category", "CollectorData")]
    public async Task Latest_collector_data_should_contain_the_expected_business_payload_for_the_seeded_contract()
    {
        var scenario = await fixture.EnsureYearlyBillingScenarioAsync();
        var options = fixture.Options;
        var collectorDataJson = scenario.CollectorDataJson;
        Assert.False(string.IsNullOrWhiteSpace(collectorDataJson));

        using var document = JsonDocument.Parse(collectorDataJson);
        var root = document.RootElement;

        Assert.Equal(options.ContractNumber, root.GetProperty("ContractNumber").GetString());
        Assert.StartsWith(options.FromDate, root.GetProperty("From").GetString(), StringComparison.Ordinal);
        Assert.StartsWith(options.UntilDate, root.GetProperty("Until").GetString(), StringComparison.Ordinal);
        Assert.True(root.GetProperty("MonthlyAmount").GetProperty("OldMonthlyAmount").GetDecimal() > 0);
        Assert.True(root.GetProperty("SaveDatabase").GetBoolean());
        Assert.False(root.GetProperty("IsErrorCollection").GetBoolean());

        var eans = root.GetProperty("EANs");
        Assert.Equal(2, eans.GetArrayLength());

        var eanValues = eans
            .EnumerateArray()
            .Select(element => element.GetProperty("Ean").GetString())
            .Where(value => !string.IsNullOrWhiteSpace(value))
            .ToHashSet(StringComparer.Ordinal);

        Assert.Contains("871692493900164636", eanValues);
        Assert.Contains("871692493900374813", eanValues);

        foreach (var ean in eans.EnumerateArray())
        {
            Assert.Equal("Activated", ean.GetProperty("Status").GetString());
            Assert.Equal("KVB", ean.GetProperty("UsageType").GetString());
            Assert.True(ean.GetProperty("Tariffs").GetArrayLength() > 0);
            Assert.True(ean.GetProperty("MasterData").GetArrayLength() > 0);
            Assert.True(ean.GetProperty("MeterReads").GetArrayLength() > 0);

            var firstTariff = ean.GetProperty("Tariffs")[0];
            Assert.False(string.IsNullOrWhiteSpace(firstTariff.GetProperty("From").GetString()));
            Assert.False(string.IsNullOrWhiteSpace(firstTariff.GetProperty("Until").GetString()));
            Assert.True(firstTariff.GetProperty("TariffType").GetInt32() > 0);

            var firstMasterData = ean.GetProperty("MasterData")[0];
            Assert.False(string.IsNullOrWhiteSpace(firstMasterData.GetProperty("GridOperatorEAN").GetString()));
            Assert.False(string.IsNullOrWhiteSpace(firstMasterData.GetProperty("MeterDeviceId").GetString()));
            Assert.False(string.IsNullOrWhiteSpace(firstMasterData.GetProperty("CapTarCode").GetString()));

            var firstMeterRead = ean.GetProperty("MeterReads")[0];
            Assert.False(string.IsNullOrWhiteSpace(firstMeterRead.GetProperty("Register").GetString()));
            Assert.True(firstMeterRead.GetProperty("Volume").ValueKind is JsonValueKind.Number);
            Assert.False(string.IsNullOrWhiteSpace(firstMeterRead.GetProperty("MeterID").GetString()));
        }

        var billingDetails = root.GetProperty("BillingDetails");
        Assert.Equal("20215441", billingDetails.GetProperty("CustomerKey").GetString());
        Assert.Equal("20215441", billingDetails.GetProperty("ClientNumber").GetString());
        Assert.Equal("customer1237@infodation.com", billingDetails.GetProperty("Email").GetString());
        Assert.Equal("Regular", billingDetails.GetProperty("OrderFlow").GetString());
        Assert.Equal(321, billingDetails.GetProperty("ResellerId").GetInt32());

        var billingAddress = root.GetProperty("BillingAddress");
        Assert.Equal("WEERT", billingAddress.GetProperty("City").GetString());
        Assert.Equal("Sint Mathiasstraat", billingAddress.GetProperty("Street").GetString());
        Assert.Equal("6004AB", billingAddress.GetProperty("Zipcode").GetString());
        Assert.Equal(1, billingAddress.GetProperty("HouseNumber").GetInt32());
    }
}

File: runners/java/src/test/java/com/kikker/billing/integrationtests/BillingInteractionLogTest.java
package com.kikker.billing.integrationtests;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.util.HashSet;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

final class BillingInteractionLogTest {
    private static final BillingFlowFixture FIXTURE = BillingFlowFixture.shared();
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    @Test
    @Tag("CollectorData")
    void latestCollectorDataShouldContainTheExpectedBusinessPayloadForTheSeededContract() throws Exception {
        BillingScenarioResult scenario = FIXTURE.ensureYearlyBillingScenario();
        BillingTestOptions options = FIXTURE.options();
        String collectorDataJson = scenario.collectorDataJson();

        assertFalse(collectorDataJson.isBlank());

        JsonNode root = OBJECT_MAPPER.readTree(collectorDataJson);
        assertEquals(options.contractNumber(), root.path("ContractNumber").asText());
        assertTrue(root.path("From").asText().startsWith(options.fromDate()));
        assertTrue(root.path("Until").asText().startsWith(options.untilDate()));
        assertTrue(root.path("MonthlyAmount").path("OldMonthlyAmount").decimalValue().doubleValue() > 0);
        assertTrue(root.path("SaveDatabase").asBoolean());
        assertFalse(root.path("IsErrorCollection").asBoolean());

        JsonNode eans = root.path("EANs");
        assertEquals(2, eans.size());

        Set<String> eanValues = new HashSet<>();
        for (JsonNode ean : eans) {
            String eanValue = ean.path("Ean").asText();
            if (!eanValue.isBlank()) {
                eanValues.add(eanValue);
            }

            assertEquals("Activated", ean.path("Status").asText());
            assertEquals("KVB", ean.path("UsageType").asText());
            assertTrue(ean.path("Tariffs").size() > 0);
            assertTrue(ean.path("MasterData").size() > 0);
            assertTrue(ean.path("MeterReads").size() > 0);

            JsonNode firstTariff = ean.path("Tariffs").get(0);
            assertNotNull(firstTariff);
            assertFalse(firstTariff.path("From").asText().isBlank());
            assertFalse(firstTariff.path("Until").asText().isBlank());
            assertTrue(firstTariff.path("TariffType").asInt() > 0);

            JsonNode firstMasterData = ean.path("MasterData").get(0);
            assertNotNull(firstMasterData);
            assertFalse(firstMasterData.path("GridOperatorEAN").asText().isBlank());
            assertFalse(firstMasterData.path("MeterDeviceId").asText().isBlank());
            assertFalse(firstMasterData.path("CapTarCode").asText().isBlank());

            JsonNode firstMeterRead = ean.path("MeterReads").get(0);
            assertNotNull(firstMeterRead);
            assertFalse(firstMeterRead.path("Register").asText().isBlank());
            assertTrue(firstMeterRead.path("Volume").isNumber());
            assertFalse(firstMeterRead.path("MeterID").asText().isBlank());
        }

        assertTrue(eanValues.contains("871692493900164636"));
        assertTrue(eanValues.contains("871692493900374813"));

        JsonNode billingDetails = root.path("BillingDetails");
        assertEquals("20215441", billingDetails.path("CustomerKey").asText());
        assertEquals("20215441", billingDetails.path("ClientNumber").asText());
        assertEquals("customer1237@infodation.com", billingDetails.path("Email").asText());
        assertEquals("Regular", billingDetails.path("OrderFlow").asText());
        assertEquals(321, billingDetails.path("ResellerId").asInt());

        JsonNode billingAddress = root.path("BillingAddress");
        assertEquals("WEERT", billingAddress.path("City").asText());
        assertEquals("Sint Mathiasstraat", billingAddress.path("Street").asText());
        assertEquals("6004AB", billingAddress.path("Zipcode").asText());
        assertEquals(1, billingAddress.path("HouseNumber").asInt());
    }
}
